File: sdks/python/agentsdk/restore.py
"""Restore — extract files from zip archives with zip-slip protection and CRC-32 verification.

Provides ``RestoreBackup`` to extract files from a zip archive created by
``CreateBackup`` into a target directory, with support for selective restore
and full integrity verification.

Security: zip-slip path traversal is detected and rejected for every entry.
Integrity: CRC-32 is verified by reading the full entry content before writing.
"""

from __future__ import annotations

import os
import zipfile
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class RestoreResult:
    """Holds the outcome of a RestoreBackup call."""

    restored: List[str] = field(default_factory=list)
    """Paths (relative to targetDir) of successfully restored files."""

    skipped: List[str] = field(default_factory=list)
    """Paths of zip entries that were skipped (e.g. not in items filter)."""


def RestoreBackup(
    zip_path: str,
    target_dir: str,
    items: Optional[List[str]] = None,
) -> RestoreResult:
    """Extract files from a zip archive into *target_dir*.

    Parameters:
        zip_path: Path to the zip archive.
        target_dir: Directory to extract files into (created if it does not exist).
        items: Optional list of entry names to restore. When ``None`` or empty,
            all entries are restored. When non-empty, only entries whose name
            exactly matches an item or has an item as a directory prefix are
            restored.

    Returns:
        A :class:`RestoreResult` with restored file paths.

    Raises:
        OSError: On the first error during extraction, aborts and returns
            the result together with any files already restored.
    """
    result = RestoreResult()

    # Open the zip archive.
    try:
        zf = zipfile.ZipFile(zip_path, "r")
    except (OSError, zipfile.BadZipFile) as e:
        raise OSError(f"restore: open zip {zip_path!r}: {e}") from e

    with zf:
        # Ensure target_dir exists.
        try:
            os.makedirs(target_dir, exist_ok=True)
        except OSError as e:
            raise OSError(
                f"restore: create target dir {target_dir!r}: {e}"
            ) from e

        # Build item lookup set for selective restore.
        selective = bool(items)
        item_set: set = set(items) if items else set()

        # Normalise target_dir to an absolute, clean path for zip-slip checks.
        try:
            abs_target = os.path.realpath(target_dir)
        except OSError as e:
            raise OSError(
                f"restore: resolve target dir {target_dir!r}: {e}"
            ) from e

        for info in zf.infolist():
            # Directory entries — create them but don't count as restored files.
            if info.is_dir():
                if selective and not _matches_item(info.filename, item_set):
                    result.skipped.append(info.filename)
                    continue
                if _is_zip_slip(info.filename, abs_target):
                    raise OSError(
                        f"restore: zip-slip detected for directory entry "
                        f"{info.filename!r}"
                    )
                dir_path = os.path.join(abs_target, info.filename)
                try:
                    os.makedirs(dir_path, exist_ok=True)
                except OSError as e:
                    raise OSError(
                        f"restore: create directory {dir_path!r}: {e}"
                    ) from e
                continue

            # Selective restore filter: skip non-matching files.
            if selective and not _matches_item(info.filename, item_set):
                result.skipped.append(info.filename)
                continue

            # Zip-slip check.
            if _is_zip_slip(info.filename, abs_target):
                raise OSError(
                    f"restore: zip-slip detected for entry {info.filename!r}"
                )

            dest_path = os.path.join(abs_target, info.filename)

            try:
                _restore_file(zf, info, dest_path)
            except OSError as e:
                raise OSError(
                    f"restore: extract {info.filename!r}: {e}"
                ) from e

            # Record relative path using forward slashes (matching zip entry convention).
            result.restored.append(info.filename)

    return result


def _matches_item(entry_name: str, item_set: set) -> bool:
    """Check whether a zip entry name should be restored during selective restore.

    An entry matches if:
    - Its name exactly equals an item (e.g. "config.json" == "config.json").
    - An item is a directory prefix of the entry (e.g. item "data/" matches
      entry "data/file.txt").
    """
    if entry_name in item_set:
        return True
    # Check directory-prefix match: item "data" should match "data/file.txt".
    for item in item_set:
        prefix = item.rstrip("/") + "/"
        if entry_name.startswith(prefix):
            return True
    return False


def _is_zip_slip(entry_path: str, target_dir: str) -> bool:
    """Return True if *entry_path* would resolve outside *target_dir* (path traversal)."""
    # Normalise both paths and ensure the resolved path is within target_dir.
    try:
        # Use os.path.normpath to collapse any ".." components.
        dest = os.path.normpath(os.path.join(target_dir, entry_path))
        abs_dest = os.path.realpath(dest)
        abs_target = os.path.realpath(target_dir)
    except OSError:
        return True  # treat resolution failure as suspicious

    # Ensure the resolved path starts with the target dir + separator.
    target_prefix = abs_target + os.sep
    if not abs_dest.startswith(target_prefix) and abs_dest != abs_target:
        return True
    return False


def _restore_file(
    zf: zipfile.ZipFile,
    info: zipfile.ZipInfo,
    dest_path: str,
) -> None:
    """Extract a single zip entry to *dest_path* with CRC-32 verification.

    Reads the full entry content first (which verifies CRC-32 via
    ``ZipFile.read``), then writes to the destination file. This mirrors the
    Go implementation's read-then-write pattern for integrity verification.
    """
    # Create parent directories.
    parent = os.path.dirname(dest_path)
    try:
        os.makedirs(parent, exist_ok=True)
    except OSError as e:
        raise OSError(f"create parent dirs: {e}") from e

    # Read the full entry content — ZipFile.read() verifies CRC-32.
    try:
        data = zf.read(info.filename)
    except (OSError, zipfile.BadZipFile) as e:
        raise OSError(f"read entry: {e}") from e

    # Write to destination file.
    try:
        with open(dest_path, "wb") as out:
            out.write(data)
    except OSError as e:
        raise OSError(f"write file: {e}") from e

File: sdks/go/restore.go
package agentsdk

import (
	"archive/zip"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

// RestoreResult holds the outcome of a RestoreBackup call.
type RestoreResult struct {
	Restored []string // Paths (relative to targetDir) of successfully restored files.
	Skipped  []string // Paths of zip entries that were skipped (e.g. not in items filter).
}

// RestoreBackup extracts files from a zip archive created by CreateBackup into
// targetDir.
//
// Parameters:
//   - zipPath: path to the zip archive.
//   - targetDir: directory to extract files into (created if it does not exist).
//   - items: optional list of entry names to restore. When nil or empty, all
//     entries are restored. When non-empty, only entries whose name exactly
//     matches an item or has an item as a directory prefix are restored.
//
// On success, returns RestoreResult with restored file paths.
// On the first error during extraction, it aborts and returns the error together
// with any files already restored.
//
// Security: zip-slip path traversal is detected and rejected.
// Integrity: archive/zip performs CRC-32 verification during decompression.
func RestoreBackup(zipPath, targetDir string, items []string) (RestoreResult, error) {
	// Open the zip archive.
	r, err := zip.OpenReader(zipPath)
	if err != nil {
		return RestoreResult{}, fmt.Errorf("restore: open zip %q: %w", zipPath, err)
	}
	defer r.Close()

	// Ensure targetDir exists.
	if err := os.MkdirAll(targetDir, 0755); err != nil {
		return RestoreResult{}, fmt.Errorf("restore: create target dir %q: %w", targetDir, err)
	}

	// Build item lookup set for selective restore.
	itemSet := make(map[string]bool, len(items))
	for _, it := range items {
		itemSet[it] = true
	}
	selective := len(itemSet) > 0

	// Normalise targetDir to an absolute, clean path for zip-slip checks.
	absTarget, err := filepath.Abs(targetDir)
	if err != nil {
		return RestoreResult{}, fmt.Errorf("restore: resolve target dir %q: %w", targetDir, err)
	}

	result := RestoreResult{
		Restored: make([]string, 0),
		Skipped:  make([]string, 0),
	}

	for _, f := range r.File {
		// Directory entries — create them but don't count as restored files.
		if f.FileInfo().IsDir() {
			if selective && !matchesItem(f.Name, itemSet) {
				result.Skipped = append(result.Skipped, f.Name)
				continue
			}
			dirPath := filepath.Join(absTarget, filepath.FromSlash(f.Name))
			if isZipSlip(f.Name, absTarget) {
				return result, fmt.Errorf("restore: zip-slip detected for directory entry %q", f.Name)
			}
			if err := os.MkdirAll(dirPath, 0755); err != nil {
				return result, fmt.Errorf("restore: create directory %q: %w", dirPath, err)
			}
			continue
		}

		// Selective restore filter: skip non-matching files.
		if selective && !matchesItem(f.Name, itemSet) {
			result.Skipped = append(result.Skipped, f.Name)
			continue
		}

		// Zip-slip check.
		if isZipSlip(f.Name, absTarget) {
			return result, fmt.Errorf("restore: zip-slip detected for entry %q", f.Name)
		}

		destPath := filepath.Join(absTarget, filepath.FromSlash(f.Name))

		if err := restoreFile(f, destPath); err != nil {
			return result, fmt.Errorf("restore: extract %q: %w", f.Name, err)
		}

		// Record relative path using forward slashes (matching zip entry convention).
		result.Restored = append(result.Restored, f.Name)
	}

	return result, nil
}

// matchesItem checks whether a zip entry name should be restored when
// selective item filtering is active.
//
// An entry matches if:
//   - Its name exactly equals an item (e.g. "config.json" == "config.json").
//   - An item is a directory prefix of the entry (e.g. item "data/" matches
//     entry "data/file.txt").
func matchesItem(entryName string, itemSet map[string]bool) bool {
	if itemSet[entryName] {
		return true
	}
	// Check directory-prefix match: item "data" should match "data/file.txt".
	// Normalise both to use forward slashes.
	for item := range itemSet {
		prefix := strings.TrimSuffix(item, "/") + "/"
		if strings.HasPrefix(entryName, prefix) {
			return true
		}
	}
	return false
}

// isZipSlip returns true if entryPath would resolve to a location outside
// targetDir after joining, indicating a path-traversal (zip-slip) attack.
func isZipSlip(entryPath, targetDir string) bool {
	// Clean the entry path and join with target dir.
	cleaned := filepath.Join(targetDir, filepath.FromSlash(entryPath))
	abs, err := filepath.Abs(cleaned)
	if err != nil {
		return true // treat resolution failure as suspicious
	}
	// Ensure the resolved path starts with the target dir.
	absTarget, err := filepath.Abs(targetDir)
	if err != nil {
		return true
	}
	return !strings.HasPrefix(abs+string(filepath.Separator), absTarget+string(filepath.Separator))
}

// restoreFile extracts a single zip file entry to destPath, creating parent
// directories as needed. Reading the file content triggers archive/zip's
// built-in CRC-32 integrity verification.
func restoreFile(zf *zip.File, destPath string) error {
	// Create parent directories.
	if err := os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
		return fmt.Errorf("create parent dirs: %w", err)
	}

	// Open the compressed file data (this validates CRC-32 on read).
	rc, err := zf.Open()
	if err != nil {
		return fmt.Errorf("open entry: %w", err)
	}
	defer rc.Close()

	// Create (or overwrite) the destination file.
	out, err := os.Create(destPath)
	if err != nil {
		return fmt.Errorf("create file: %w", err)
	}
	defer out.Close()

	if _, err := io.Copy(out, rc); err != nil {
		return fmt.Errorf("write file: %w", err)
	}

	return nil
}
